### PR TITLE
Refactor partition

### PR DESCRIPTION
**Pull request by eedalong**
#112 
- [x] add `quiver_partition` for partition graph topology
- [x] add `load_quiver_partition` for loading partition result
- [x] add unit test for different partition algorithm
- [x] partition result is aligned with original version implemented by Tzy @ZenoTan

### CODE REVIEW

File: srcs/python/quiver/feature.py [hunk 1]
```diff
@@ -217,12 +198,10 @@ def from_cpu_tensor(self, cpu_tensor: torch.Tensor):
             cpu_tensor (torch.FloatTensor): input cpu tensor
         """
         if self.cache_policy == "device_replicate":
-            cache_memory_budget = self.cal_memory_budget_bytes(
-                self.device_cache_size)
+            cache_memory_budget = cal_memory_budget_to_bytes(self.device_cache_size)
```
Comment (lgarithm) [inline srcs/python/quiver/feature.py:201]: `cal_memory_budget_to_bytes(self.device_cache_size)` should be called once on construction.

File: srcs/python/quiver/partition.py [hunk 2]
```diff
@@ -1,127 +1,139 @@
...
+    device = torch.cuda.current_device()
+    partitioned_num = len(probs)
+
+    probs = [prob.to(device) for prob in probs]
+    total_node_num = probs[0].size(0)
+
+    res = [[] for _ in range(partitioned_num)]
+
+    chunk_size = QUIVER_MAGIC_NUMBER * partitioned_num
+    chunk_num = (total_node_num + chunk_size - 1) // chunk_size
```
Comment (ZenoTan) [inline srcs/python/quiver/partition.py:31]: I think this number could be used as an argument but with default value:
`partition_without_replication(probs, chunk_num=64)`

File: srcs/python/quiver/utils.py [hunk 3]
```diff
@@ -255,3 +255,28 @@ def init_p2p(device_list: List[int]):
         device_list (List[int]): device list
     """
     torch_qv.init_p2p(device_list)
+
+
+def cal_memory_budget_to_bytes(memory_budget):
```
[the remark this answers is not in the document]
Comment (eedalong) [reply srcs/python/quiver/utils.py:260]: Beautiful

File: tests/python/cuda/test_partition.py [hunk 4]
```diff
@@ -1,228 +1,392 @@
...
-    train_idx1 = train_idx1.to(0)
-
-    total = torch.zeros((nodes, ), device=0)
-    remote = torch.ones((nodes, ), device=0)
+    quiver_sampler = quiver.pyg.GraphSageSampler(csr_topo, [15, 10, 5], 0, mode="UVA")
+    print(f"average degree of paper100M = {torch.sum(csr_topo.degree) / csr_topo.node_count}")
+    return train_idx, csr_topo, quiver_sampler
+
+def load_products():
+    root = "/home/dalong/data/products/"
```
[the remark this answers is not in the document]
Comment (eedalong) [reply tests/python/cuda/test_partition.py:168]: My bad

File: srcs/python/quiver/partition.py (hunk 2 above)
Comment (eedalong) [reply srcs/python/quiver/partition.py:33]: okk~

File: srcs/python/quiver/feature.py (hunk 1 above)
Comment (eedalong) [reply srcs/python/quiver/feature.py]: We only calculate once for this parse for each quiver.Feature

**Review by luomai — APPROVED** [review]

**Pull request merged**